trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/rest/ParsedRequestBuilder.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.rest;
+
+import com.google.common.base.Splitter;
+import jakarta.ws.rs.core.MultivaluedMap;
+import jakarta.ws.rs.core.UriBuilder;
+import org.glassfish.jersey.server.ContainerRequest;
+
+import java.io.InputStream;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static io.trino.s3.proxy.server.collections.MultiMapHelper.lowercase;
+
+class ParsedRequestBuilder
+{
+    private ParsedRequestBuilder() {}
+
+    static ParsedRequest fromRequest(ContainerRequest request)
+    {
+        return new ParsedRequest(request.getRequestUri(), request.getRequestHeaders(), request.getUriInfo().getQueryParameters(), request.getMethod());
+    }
+
+    static ParsedS3Request fromRequest(String requestPath, ContainerRequest request, Optional<String> serverHostName)
```

Comment:
I feel that this distinction is not made clear. One may call `fromRequest` and obtain:
- a `ParsedS3Request` with lowercased headers, or
- a `ParsedRequest` with no modifications

Should one of these records inherit from the other?

Reply:
At one point I had a common `interface`. The issue is that they have different meanings. `ParsedRequest` is not the best name, I agree. Maybe `EmulatedRequest`?

Reply:
@vagaerg I renamed `ParseRequest` to simply `Request`. I realize it's a common name but Airlift does it and Trino does it for things like Query. This way we have a Trino Proxy representation of a "request" as well as a "ParsedS3Request". Also, in ParsedS3Request I renamed requestHeaders to lowercaseHeaders. I still think we should have our own Multimap that abstracts common behavior.

Reply:
Agreed on the multimap - see #61 

This sounds reasonable for now, I wonder whether we could turn a `ContainerRequest` into a `Request`, and then optionally use that plus some metadata to build a `ParsedS3Request`.

That way we only use `ContainerRequest` at the edge, when we first get the request, then we only deal with our representation of it.

I'll have a look when implementing #61